### PR TITLE
AARCH64 SHA512: implementation using crypto instructions added

### PR DESCRIPTION
**Pull request by SparkiDev**
Checks system register for the feature before using the SHA512
instructions.
Added SHA512 input data alignment test.

### CODE REVIEW

**Review by ejohnstown — CHANGES_REQUESTED** [review]
1. When I build with armasm-sha512 enabled, it isn't any faster. Granted, the M1 might not have the feature available.
2. When I build with armasm enabled, the linker can't find SHA2-512 functions for the other output sizes. (wc_InitSha512_224, etc)

**Comment by SparkiDev** [comment]
Rebased.
M1 should have the instructions - it is meant to be ARMV8.4-A.
Please check whether it detects availability of the instructions correctly. (See if it exists even when it doesn't detect it.)

**Comment by ejohnstown** [comment]
> Rebased.
> M1 should have the instructions - it is meant to be ARMV8.4-A.
> Please check whether it detects availability of the instructions correctly. (See if it exists even when it doesn't detect it.)

I forced the testwolfcrypt test to run the cpuid_get_flags() function. Apparently, it isn't normally present or it isn't getting executed. When I call it explicitly, the test crashes with "illegal hardware instruction". Through the debugger, this happens on the `MRS %[r] ID_AA64ISAR0_EL1` instruction in aarch64_sha512_instructions().

**Comment by SparkiDev** [comment]
Don't try to detect anymore.
If you use the configuration item then you will use the SHA512 instruction version.

**Comment by ejohnstown** [comment]
I'm beginning to think the M1 doesn't support sha512.

```
  CPPAS    wolfcrypt/src/port/arm/src_libwolfssl_la-armv8-sha512-asm.lo
wolfcrypt/src/port/arm/armv8-sha512-asm.S:1256:2: error: instruction requires: sha3
 sha512h q27, q21, v22.2d
 ^
wolfcrypt/src/port/arm/armv8-sha512-asm.S:1258:2: error: instruction requires: sha3
 sha512h2 q27, q25, v24.2d
 ^
...
```

So, I won't hold this up because of the M1. Let me know if you want to hack on this more and I can give you a login on my Mac you can SSH into.

**Review by ejohnstown — APPROVED** [review]

**Comment by SparkiDev** [comment]
I'm done with this now. Since there is a configure option for it, then customers can decide for themselves whether to use it.

File: configure.ac [hunk 1]
```diff
@@ -1478,6 +1478,17 @@ then
 fi
 
 
+AC_ARG_ENABLE([armasm-sha512],
```
Comment (dgarske) [inline configure.ac:1481]: Perhaps thinking long term for possible additional ARMv8.2 SHA instruction support (some of which are optional), would it make sense to support something like `--enable-armasm=sha512,sha3`?

* ARMv8.2-SHA SHA-3 instructions (EOR3, RAX1, XAR, BCAX) -> `WOLFSSL_ARMASM_CRYPTO_SHA3`
* ARMv8.2-SHA SHA512 instructions (SHA512) -> `WOLFSSL_ARMASM_CRYPTO_SHA512`

**Review by dgarske — CHANGES_REQUESTED** [review]

File: configure.ac [hunk 2]
```diff
@@ -1484,8 +1484,25 @@ AC_ARG_ENABLE([armasm],
     [ ENABLED_ARMASM=$enableval ],
     [ ENABLED_ARMASM=no ]
     )
-if test "$ENABLED_ARMASM" = "yes" && test "$ENABLED_ASM" = "yes"
-then
+if test "$ENABLED_ARMASM" != "no" && test "$ENABLED_ASM" = "yes"
```
Comment (dgarske) [inline configure.ac:1487]: Can you update the help to reflect this new option? Also the comment above says “AES ARM”, can you fix that?

**Review by dgarske — CHANGES_REQUESTED** [review]

File: configure.ac (hunk 2 above)
Comment (SparkiDev) [reply configure.ac:1487]: Done.
Let me know if there is anything else.

**Review by dgarske — APPROVED** [review]

**Pull request merged**